pkg/cap/cluster/host.go

```diff
@@ -0,0 +1,225 @@
+package cluster
+
+import (
+	"context"
+	"sync"
+
+	"capnproto.org/go/capnp/v3"
+	"capnproto.org/go/capnp/v3/rpc"
+	"capnproto.org/go/capnp/v3/server"
+	"github.com/libp2p/go-libp2p-core/peer"
+	api "github.com/wetware/ww/internal/api/cluster"
+	"github.com/wetware/ww/pkg/vat"
+)
+
+var (
+	HostCapability = vat.BasicCap{
+		"hostAnchor/packed",
+		"hostAnchor"}
+	HostDefaultPolicy = server.Policy{
+		// HACK:  raise MaxConcurrentCalls to mitigate known deadlock condition.
+		//        https://github.com/capnproto/go-capnproto2/issues/189
+		MaxConcurrentCalls: 64,
+		AnswerQueueSize:    64,
+	}
+)
+
+type HostAnchor struct {
+	Peer peer.ID
+	Vat  vat.Network
+
+	once   sync.Once
+	client api.Host
+}
+
+func (ha HostAnchor) Path() []string {
+	return []string{ha.Peer.String()}
+}
+
+func (ha HostAnchor) Ls(ctx context.Context) (AnchorIterator, error) {
+	if err := ha.bootstrapOnce(ctx); err != nil {
+		return nil, err
+	}
+
+	fut, release := ha.client.Ls(ctx, func(a api.Anchor_ls_Params) error {
+		return nil
+	})
+	select {
+	case <-fut.Done():
+		results, err := fut.Struct()
+		if err != nil {
+			return nil, err
+		}
+		children, err := results.Children()
+		if err != nil {
+			return nil, err
+		} else {
+			return &ContainerAnchorIterator{path: ha.Path(), children: children, release: release}, err
+		}
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (ha HostAnchor) Walk(ctx context.Context, path []string) (Anchor, error) {
+	if err := ha.bootstrapOnce(ctx); err != nil {
+		return nil, err
+	}
+
+	fut, release := ha.client.Walk(ctx, func(a api.Anchor_walk_Params) error {
+		capPath, err := a.NewPath(int32(len(path)))
+		if err != nil {
+			return err
+		}
+		for i, e := range path {
+			if err := capPath.Set(i, e); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+
+	return ContainerAnchor{path: append(ha.Path(), path...), client: api.Container(fut.Anchor()), release: release}, nil
+}
+
+func (ha HostAnchor) bootstrapOnce(ctx context.Context) error {
+	var (
+		conn *rpc.Conn
+		err  error
+	)
+
+	if ha.client.Client != nil {
+		return nil
+	}
+
+	ha.once.Do(func() {
+		conn, err = ha.Vat.Connect(
+			ctx,
+			peer.AddrInfo{ID: ha.Peer},
+			HostCapability,
+		)
+		if err == nil {
+			ha.client = api.Host{Client: conn.Bootstrap(ctx)}
+		}
+	})
+
+	return err
+}
+
+type HostAnchorIterator struct {
+	Vat     vat.Network
+	It      *Iterator
+	Release capnp.ReleaseFunc
+}
+
+func (it HostAnchorIterator) Next(ctx context.Context) bool {
+	return it.It.Next(ctx)
+}
+
+func (it HostAnchorIterator) Finish() {
+	it.Release()
+}
+
+func (it HostAnchorIterator) Anchor() Anchor {
+	return HostAnchor{
+		Peer: it.It.Record().Peer(),
```

Comment:
`HostAnchor` should contain a `peer.AddrInfo` instead of just the `peer.ID`.  This allows us to eliminate an (expensive!) DHT call when we have a full record, as in the current case.

Reply:
The thing is that we don't have a full record. The record is the capnp Record struct:
```
struct Record {
        peer @0 :Text;
        ttl @1 :Int64;
        seq @2 :UInt64;
    }
```
Moreover, in order to make the `Record struct` to contain an `AddrInfo`, we would need to change the `RoutingTable` interface.